File: crates/server/src/routes/kanban.rs
//! API routes for the kanban task board.
//!
//! Provides endpoints for listing tasks, viewing the board, and getting stats.
//! Task mutations (create, claim, update, review) happen through agent tools.

use axum::{
    extract::{Path, Query, State},
    Json,
};
use chrono::Utc;
use serde::{Deserialize, Serialize};

use rushdino_agent::kanban_store::{KanbanBoardStats, KanbanTask, TaskStatus};
use rushdino_common::Result;

use crate::state::AppState;

// ---------------------------------------------------------------------------
// Query params
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
pub struct KanbanListQuery {
    /// Filter by status (e.g. "backlog", "in_progress", "done").
    pub status: Option<String>,
    /// Filter by assigned agent name.
    pub agent: Option<String>,
    /// Filter by source request ID.
    pub source: Option<String>,
}

// ---------------------------------------------------------------------------
// Response types
// ---------------------------------------------------------------------------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct KanbanBoardResponse {
    pub generated_at: String,
    pub stats: KanbanBoardStats,
    pub columns: KanbanBoardColumns,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct KanbanBoardColumns {
    pub backlog: Vec<KanbanTask>,
    pub claimed: Vec<KanbanTask>,
    pub in_progress: Vec<KanbanTask>,
    pub blocked: Vec<KanbanTask>,
    pub in_review: Vec<KanbanTask>,
    pub done: Vec<KanbanTask>,
    pub failed: Vec<KanbanTask>,
}

// ---------------------------------------------------------------------------
// Handlers
// ---------------------------------------------------------------------------

/// GET /api/kanban/board — Full board view with all columns and stats.
pub async fn get_kanban_board(State(state): State<AppState>) -> Result<Json<KanbanBoardResponse>> {
    let engine = state.engine()?;
    let store = engine.kanban_store();

    let stats = store.get_board_stats().await?;
    let all_tasks = store.list_tasks(None).await?;

    let mut columns = KanbanBoardColumns {
        backlog: Vec::new(),
        claimed: Vec::new(),
        in_progress: Vec::new(),
        blocked: Vec::new(),
        in_review: Vec::new(),
        done: Vec::new(),
        failed: Vec::new(),
    };

    for task in all_tasks {
        match task.status {
            TaskStatus::Backlog => columns.backlog.push(task),
            TaskStatus::Claimed => columns.claimed.push(task),
            TaskStatus::InProgress => columns.in_progress.push(task),
            TaskStatus::Blocked => columns.blocked.push(task),
            TaskStatus::InReview => columns.in_review.push(task),
            TaskStatus::Done => columns.done.push(task),
            TaskStatus::Failed => columns.failed.push(task),
        }
    }

    Ok(Json(KanbanBoardResponse {
        generated_at: Utc::now().to_rfc3339(),
        stats,
        columns,
    }))
}

/// GET /api/kanban/tasks — List tasks with optional filters.
pub async fn list_kanban_tasks(
    State(state): State<AppState>,
    Query(query): Query<KanbanListQuery>,
) -> Result<Json<Vec<KanbanTask>>> {
    let engine = state.engine()?;
    let store = engine.kanban_store();

    let tasks = if let Some(ref agent) = query.agent {
        store.list_tasks_by_agent(agent).await?
    } else if let Some(ref source) = query.source {
        store.list_tasks_by_source(source).await?
    } else if let Some(ref status) = query.status {
        let status = TaskStatus::from_str_loose(status);
        store.list_tasks(Some(status)).await?
    } else {
        store.list_tasks(None).await?
    };

    Ok(Json(tasks))
}

/// GET /api/kanban/tasks/:id — Get a single task by ID.
pub async fn get_kanban_task(
    State(state): State<AppState>,
    Path(task_id): Path<String>,
) -> Result<Json<KanbanTask>> {
    let engine = state.engine()?;
    let store = engine.kanban_store();
    let task = store.get_task(&task_id).await?;
    Ok(Json(task))
}

/// DELETE /api/kanban/tasks/:id — Delete a task and its subtasks.
pub async fn delete_kanban_task(
    State(state): State<AppState>,
    Path(task_id): Path<String>,
) -> Result<Json<serde_json::Value>> {
    let engine = state.engine()?;
    let store = engine.kanban_store();
    store.delete_task(&task_id).await?;
    Ok(Json(serde_json::json!({ "deleted": task_id })))
}

/// GET /api/kanban/stats — Board statistics only.
pub async fn get_kanban_stats(State(state): State<AppState>) -> Result<Json<KanbanBoardStats>> {
    let engine = state.engine()?;
    let store = engine.kanban_store();
    let stats = store.get_board_stats().await?;
    Ok(Json(stats))
}

File: crates/desktop-app/ui/src/api/kanban.ts
import { apiFetch } from './bootstrap'

/* Mirrors crates/agent/src/kanban_store.rs types (serialized as camelCase
 * via serde rename on the KanbanBoardResponse; the task struct itself is
 * snake_case because it isn't rename-cased on the Rust side). */

export type TaskStatus =
  | 'backlog'
  | 'claimed'
  | 'in_progress'
  | 'blocked'
  | 'in_review'
  | 'done'
  | 'failed'

export type TaskPriority = 'low' | 'medium' | 'high' | 'critical'

export type KanbanTask = {
  id: string
  source_request_id?: string | null
  parent_task_id?: string | null
  title: string
  description: string
  tags: string[]
  priority: TaskPriority
  status: TaskStatus
  assigned_agent?: string | null
  conversation_id?: string | null
  result?: string | null
  review_feedback?: string | null
  block_reason?: string | null
  complexity_level?: number
  depth?: number
  created_at: string
  updated_at: string
  claimed_at?: string | null
  completed_at?: string | null
  revision_count?: number
}

export type KanbanBoardStats = {
  total: number
  backlog: number
  claimed: number
  in_progress: number
  blocked: number
  in_review: number
  done: number
  failed: number
}

export type KanbanBoardColumns = {
  backlog: KanbanTask[]
  claimed: KanbanTask[]
  inProgress: KanbanTask[]
  blocked: KanbanTask[]
  inReview: KanbanTask[]
  done: KanbanTask[]
  failed: KanbanTask[]
}

export type KanbanBoardResponse = {
  generatedAt: string
  stats: KanbanBoardStats
  columns: KanbanBoardColumns
}

export async function getBoard(): Promise<KanbanBoardResponse> {
  const res = await apiFetch('/api/kanban/board')
  if (!res.ok) throw new Error(`kanban.board: ${res.status}`)
  return (await res.json()) as KanbanBoardResponse
}

export async function listTasks(filter?: {
  status?: TaskStatus
  agent?: string
  source?: string
}): Promise<KanbanTask[]> {
  const qs = new URLSearchParams()
  if (filter?.status) qs.set('status', filter.status)
  if (filter?.agent) qs.set('agent', filter.agent)
  if (filter?.source) qs.set('source', filter.source)
  const q = qs.toString()
  const res = await apiFetch(`/api/kanban/tasks${q ? `?${q}` : ''}`)
  if (!res.ok) throw new Error(`kanban.tasks: ${res.status}`)
  return (await res.json()) as KanbanTask[]
}

export async function deleteTask(id: string): Promise<void> {
  const res = await apiFetch(`/api/kanban/tasks/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
  if (!res.ok && res.status !== 404) {
    throw new Error(`kanban.delete: ${res.status}`)
  }
}
